File: backend/src/scrapers/gdelt.ts
import { EventCreate } from "../models/types";
import { bulkUpsert } from "../services/eventService";
import { broadcast } from "../services/broadcast";

const GDELT_LAST_UPDATE_URL = "http://data.gdeltproject.org/gdeltv2/lastupdate.txt";

// CAMEO root codes for conflict events
const CONFLICT_ROOT_CODES = new Set(["14", "17", "18", "19", "20"]);

const CAMEO_TYPE_MAP: Record<string, string> = {
  "14": "protest",
  "17": "coercion",
  "18": "assault",
  "19": "fight",
  "20": "mass_violence",
};

// GDELT CSV column indexes (tab-separated, no header)
const COL = {
  GLOBAL_EVENT_ID: 0,
  DAY: 1,
  ACTOR1_NAME: 6,
  ACTOR2_NAME: 16,
  EVENT_ROOT_CODE: 26,
  EVENT_CODE: 27,
  ACTION_GEO_FULLNAME: 49,
  ACTION_GEO_COUNTRY: 50,
  ACTION_GEO_LAT: 53,
  ACTION_GEO_LONG: 54,
  SOURCE_URL: 57,
};

export async function scrapeGdelt(): Promise<number> {
  console.log("[gdelt] Starting scrape...");

  // Get latest update URL
  const updateResp = await fetch(GDELT_LAST_UPDATE_URL);
  const updateText = await updateResp.text();

  let exportUrl: string | null = null;
  for (const line of updateText.trim().split("\n")) {
    const parts = line.trim().split(" ");
    if (parts.length >= 3 && parts[2].endsWith(".export.CSV.zip")) {
      exportUrl = parts[2];
      break;
    }
  }

  if (!exportUrl) {
    console.error("[gdelt] Could not find export CSV URL");
    return 0;
  }

  console.log(`[gdelt] Downloading ${exportUrl}`);
  const csvResp = await fetch(exportUrl);
  const zipBuffer = await csvResp.arrayBuffer();

  // Decompress ZIP — use the built-in DecompressionStream approach
  // Since we're dealing with a ZIP file, we need a proper unzip
  const { Readable } = await import("stream");
  const { createUnzip } = await import("zlib");
  const { Parse } = await import("unzipper" as any).catch(() => null) || {};

  // Simpler approach: use AdmZip-style parsing or raw approach
  // Let's use a buffer-based approach
  let csvText: string;
  try {
    const AdmZip = (await import("adm-zip")).default;
    const zip = new AdmZip(Buffer.from(zipBuffer));
    const entries = zip.getEntries();
    csvText = entries[0].getData().toString("utf-8");
  } catch {
    // Fallback: try importing unzipper
    console.error("[gdelt] Could not unzip CSV. Install adm-zip: npm i adm-zip");
    return 0;
  }

  const lines = csvText.split("\n");
  const events: EventCreate[] = [];

  for (const line of lines) {
    if (!line.trim()) continue;
    const cols = line.split("\t");
    if (cols.length < 58) continue;

    const rootCode = cols[COL.EVENT_ROOT_CODE];
    if (!CONFLICT_ROOT_CODES.has(rootCode)) continue;

    const lat = parseFloat(cols[COL.ACTION_GEO_LAT]);
    const lon = parseFloat(cols[COL.ACTION_GEO_LONG]);
    if (isNaN(lat) || isNaN(lon) || (lat === 0 && lon === 0)) continue;

    const eventType = CAMEO_TYPE_MAP[rootCode] || "conflict";
    const dayStr = cols[COL.DAY];
    const eventDate = new Date(
      `${dayStr.slice(0, 4)}-${dayStr.slice(4, 6)}-${dayStr.slice(6, 8)}`
    );
    if (isNaN(eventDate.getTime())) continue;

    const actors: string[] = [];
    const actor1 = cols[COL.ACTOR1_NAME]?.trim();
    const actor2 = cols[COL.ACTOR2_NAME]?.trim();
    if (actor1) actors.push(actor1);
    if (actor2) actors.push(actor2);

    const locationName = cols[COL.ACTION_GEO_FULLNAME]?.trim() || undefined;
    const country = cols[COL.ACTION_GEO_COUNTRY]?.trim() || undefined;
    const sourceUrl = cols[COL.SOURCE_URL]?.trim() || undefined;

    const actorDesc =
      actor1 && actor2 ? ` between ${actor1} and ${actor2}` : "";
    const title =
      `${eventType.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase())}${actorDesc} in ${locationName || "Unknown"}`.slice(
        0,
        512
      );

    events.push({
      source: "gdelt",
      source_id: cols[COL.GLOBAL_EVENT_ID],
      event_type: eventType,
      sub_event_type: cols[COL.EVENT_CODE] || undefined,
      title,
      url: sourceUrl,
      latitude: lat,
      longitude: lon,
      location_name: locationName,
      country,
      event_date: eventDate,
      fatalities: 0,
      actors: actors.length > 0 ? actors : undefined,
      tags: [`cameo_${cols[COL.EVENT_CODE]}`],
    });
  }

  console.log(`[gdelt] Parsed ${events.length} conflict events from ${lines.length} total rows`);
  const newCount = await bulkUpsert(events);
  console.log(`[gdelt] Inserted ${newCount} new events`);

  if (newCount > 0) {
    broadcast({ type: "scraper_update", source: "gdelt", new_events: newCount });
  }

  return newCount;
}

File: backend/app/scrapers/gdelt.py
import csv
import io
import logging
import zipfile
from datetime import datetime

import httpx

from app.schemas.event import EventCreate
from app.scrapers.base import BaseScraper

logger = logging.getLogger(__name__)

GDELT_LAST_UPDATE_URL = "http://data.gdeltproject.org/gdeltv2/lastupdate.txt"

# CAMEO root codes for conflict-related events
CONFLICT_ROOT_CODES = {"14", "17", "18", "19", "20"}

# CAMEO root code → event type mapping
CAMEO_TYPE_MAP = {
    "14": "protest",
    "17": "coercion",
    "18": "assault",
    "19": "fight",
    "20": "mass_violence",
}

# GDELT export CSV columns (fixed positions, no header)
COL_GLOBAL_EVENT_ID = 0
COL_DAY = 1
COL_ACTOR1_NAME = 6
COL_ACTOR2_NAME = 16
COL_EVENT_ROOT_CODE = 26
COL_EVENT_CODE = 27
COL_NUM_MENTIONS = 31
COL_NUM_SOURCES = 32
COL_AVG_TONE = 34
COL_ACTOR1_GEO_COUNTRY = 37
COL_ACTOR1_GEO_LAT = 39
COL_ACTOR1_GEO_LONG = 40
COL_ACTOR2_GEO_COUNTRY = 44
COL_ACTION_GEO_FULLNAME = 49
COL_ACTION_GEO_COUNTRY = 50
COL_ACTION_GEO_LAT = 53
COL_ACTION_GEO_LONG = 54
COL_SOURCE_URL = 57


class GdeltScraper(BaseScraper):
    source_name = "gdelt"

    async def fetch_raw(self) -> list[dict]:
        async with httpx.AsyncClient(timeout=60) as client:
            # Get the latest update file URL
            resp = await client.get(GDELT_LAST_UPDATE_URL)
            resp.raise_for_status()

            # First line contains the export CSV URL
            lines = resp.text.strip().split("\n")
            export_url = None
            for line in lines:
                parts = line.strip().split(" ")
                if len(parts) >= 3 and parts[2].endswith(".export.CSV.zip"):
                    export_url = parts[2]
                    break

            if not export_url:
                logger.error("[gdelt] Could not find export CSV URL")
                return []

            logger.info(f"[gdelt] Downloading {export_url}")
            resp = await client.get(export_url)
            resp.raise_for_status()

            # Unzip in memory and parse CSV
            with zipfile.ZipFile(io.BytesIO(resp.content)) as zf:
                csv_filename = zf.namelist()[0]
                with zf.open(csv_filename) as f:
                    text = f.read().decode("utf-8", errors="replace")

            rows = []
            reader = csv.reader(io.StringIO(text), delimiter="\t")
            for row in reader:
                if len(row) < 58:
                    continue

                root_code = row[COL_EVENT_ROOT_CODE]
                if root_code not in CONFLICT_ROOT_CODES:
                    continue

                lat = row[COL_ACTION_GEO_LAT].strip()
                lon = row[COL_ACTION_GEO_LONG].strip()
                if not lat or not lon:
                    continue

                rows.append(row)

            return rows

    def normalize(self, raw: list) -> EventCreate | None:
        try:
            lat = float(raw[COL_ACTION_GEO_LAT])
            lon = float(raw[COL_ACTION_GEO_LONG])
        except (ValueError, TypeError):
            return None

        if lat == 0 and lon == 0:
            return None

        root_code = raw[COL_EVENT_ROOT_CODE]
        event_type = CAMEO_TYPE_MAP.get(root_code, "conflict")

        day_str = raw[COL_DAY]
        try:
            event_date = datetime.strptime(day_str, "%Y%m%d")
        except ValueError:
            return None

        actors = []
        actor1 = raw[COL_ACTOR1_NAME].strip()
        actor2 = raw[COL_ACTOR2_NAME].strip()
        if actor1:
            actors.append(actor1)
        if actor2:
            actors.append(actor2)

        location_name = raw[COL_ACTION_GEO_FULLNAME].strip() or None
        country_code = raw[COL_ACTION_GEO_COUNTRY].strip() or None

        source_url = raw[COL_SOURCE_URL].strip() if len(raw) > COL_SOURCE_URL else None

        actor_desc = f" between {actor1} and {actor2}" if actor1 and actor2 else ""
        title = f"{event_type.replace('_', ' ').title()}{actor_desc} in {location_name or 'Unknown'}"

        return EventCreate(
            source="gdelt",
            source_id=raw[COL_GLOBAL_EVENT_ID],
            event_type=event_type,
            sub_event_type=raw[COL_EVENT_CODE],
            title=title[:512],
            description=None,
            url=source_url,
            latitude=lat,
            longitude=lon,
            location_name=location_name,
            country=country_code,
            region=None,
            event_date=event_date,
            fatalities=0,
            actors=actors if actors else None,
            tags=[f"cameo_{raw[COL_EVENT_CODE]}"] if raw[COL_EVENT_CODE] else None,
        )
